### PR TITLE
For input formats with column names, search columns case-insensitively if there is no ambiguity

### PR DESCRIPTION
**Pull request by Manerone**
### Changelog category
- Improvement

### Changelog entry

Adds a new setting `input_format_column_name_matching_mode` which allows different case sensitivities for input formats. 

### Documentation entry for user-facing changes

The possible options for the new setting are:

- `match_case`: will do a case-sensitive search of the input columns.
- `ignore_case`: will do a case-insensitive search of the input columns.
- `auto`: will first try to match case-sensitive, if it fails tries to match case-insensitive.

NOTE: case-insensitive searches can have ambiguity (e.g. Name and namE are ambiguous). If any ambiguity is detected, an exception will be thrown. 

Supported formats:

- JSONEachRow
- CSVWithNames
- JSONColumns
- BSONEachRow
- RowBinaryWithNames

#### Match Case

```
CREATE TABLE test (id Int, name String);
SET input_format_column_name_matching_mode = 'match_case'
INSERT INTO test FORMAT JSONEachRow {"id": 0, "NAME": "aaa"};

SELECT * FROM test
┌─id─┬─name─┐
│  0 │      │
└────┴──────┘
```

#### Ignore Case

```
CREATE TABLE test (id Int, name String);
SET input_format_column_name_matching_mode = 'ignore_case'
INSERT INTO test FORMAT JSONEachRow {"id": 0, "NAME": "aaa"};

SELECT * FROM test
┌─id─┬─name─┐
│  0 │ aaa  │
└────┴──────┘

-- Ambiguity
CREATE TABLE test2 (id Int, ID Int);
SET input_format_column_name_matching_mode = 'ignore_case'
INSERT INTO test FORMAT JSONEachRow {"id": 0}

DB::Exception: Ambiguous field (`id`) when processing data: (at row 1)

-- Ambiguity is only detected if the columns are referenced
CREATE TABLE test(id Int, ID Int, name String)
SET input_format_column_name_matching_mode = 'ignore_case'
INSERT INTO test FORMAT JSONEachRow {"name": "aa"}

SELECT * FROM test

┌─id─┬─ID─┬─name─┐
│  0 │  0 │ aa   │
└────┴────┴──────┘


```

#### Auto Case

```
CREATE TABLE test (id Int, ID Int);
SET input_format_column_name_matching_mode = 'auto'
INSERT INTO test FORMAT JSONEachRow {"id": 1}

SELECT * FROM test

┌─id─┬─ID─┐
│  1 │  0 │
└────┴────┘

-- Ambiguity
CREATE TABLE test (id Int, ID Int);
SET input_format_column_name_matching_mode = 'auto'
INSERT INTO test FORMAT JSONEachRow {"Id": 1} -- Notice that `Id` doesn't match any column, so it is matched by a case insensitive search

DB::Exception: Ambiguous field (`id`) when processing data: (at row 1)
```
### Dicussion - possible corner case

```
CREATE TABLE test(id Int);

SET input_format_column_name_matching_mode = 'auto'

INSERT INTO json_test FORMAT JSONEachRow {"ID": 444, "id": 123};

SELECT * FROM json_test

   ┌──id─┐
1. │ 444 │
   └─────┘
```

Influence of flag`input_format_json_ignore_unnecessary_fields`. If the flag is set to `true`, we set the case match to `auto`, then the reader will match `ID` with the table field `id` and it will consider the rest of the columns irrelevant. We can, of course, still check the remaining columns, but it would go against a previous comment found in the code:

```
            // Keep parsing the remaining fields in case of the json is invalid.
            // But not look up the name in the name_map since the cost cannot be ignored
```

The ambiguity check works fine in the following scenario:
```
CREATE TABLE test(id Int, name String, age Int);

SET input_format_column_name_matching_mode = 'auto'

INSERT INTO json_test FORMAT JSONEachRow {"ID": 444, "id": 123};

SELECT * FROM json_test

DB::Exception: Ambiguous field (`id`) when processing data: (at row 1)

```

Here it will work because total_columns is 3, and the ambiguity happens before reaching the total_columns. 

---

> [!NOTE]
> **Medium Risk**
> Changes column-name lookup behavior across multiple input formats and adds new ambiguity/duplicate-field errors, which could alter ingestion outcomes for existing pipelines relying on previous matching/skip behavior.
> 
> **Overview**
> Adds a new setting, `input_format_with_names_case_insensitive_column_matching`, to control how input formats that carry column names match those names to table columns: *case-sensitive*, *case-insensitive*, or *AUTO* (exact match first, then case-insensitive).
> 
> Introduces `CaseAwareBlockNameMap` to implement this matching (including ambiguity detection when multiple table columns collide under case-insensitive comparison) and wires it into `JSONEachRow`, `JSONColumns`, `BSONEachRow`, and `RowBinaryWithNames`/`CSVWithNames` header mapping. Parsing now throws on ambiguous mappings and on repeated fields that resolve to the same target column, with new stateless tests covering normal and ambiguous cases.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 77a9311b5a6a6e8da1289eb69c941ec65735ec67. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### Version info
- Merged into: `26.4.1.1102`

### CODE REVIEW

File: src/Core/CaseAwareBlockNameMap.cpp [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+        if (idx != CaseAwareBlockNameMap::NOT_FOUND)
+        {
+            return idx;
+        }
+
+        // Check if the key has a match ignoring case
+        return i_map.get(key);
+    }
+
+    bool stringCompare(std::string_view left, std::string_view right) const { return map.stringCompare(left, right); }
```
Comment (Avogar) [inline src/Core/CaseAwareBlockNameMap.cpp]: Should we use `map.stringCompare(left, right) || i_map.stringCompare(left, right)` here? If no, why?

File: src/Core/CaseAwareBlockNameMap.cpp [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+}
+
+struct CaseInsensitiveHash
+{
+    size_t operator()(const std::string_view key) const
+    {
+        // TODO: there is probably a way to calculate this without copying the string
+        std::string key_s{key};
+        Poco::toLowerInPlace(key_s);
+        return sipHash64(key_s);
```
Comment (Avogar) [inline src/Core/CaseAwareBlockNameMap.cpp]:
```suggestion
    SipHash hash;
    for (char c : key)
        hash.update(static_cast<unsigned char>(std::tolower(static_cast<unsigned char>(c))));
    return hash.get64();
```
Something like this can work

File: src/Core/CaseAwareBlockNameMap.cpp [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+
+struct CaseInsensitiveEquality
+{
+    bool operator()(const std::string_view left, const std::string_view right) const { return CaseInsensitiveEquality::equal(left, right); }
+
+    static bool equal(const std::string_view left, const std::string_view right)
+    {
+        if (left.size() != right.size())
+            return false;
+        return Poco::toLower(std::string(left)) == Poco::toLower(std::string(right));
```
Comment (Avogar) [inline src/Core/CaseAwareBlockNameMap.cpp]:
```suggestion
        return strncasecmp(left.data(), right.data(), left.size()) == 0;;
```
I believe this would work and we can avoid creating strings from string_view.

File: src/Formats/ColumnMapping.cpp [hunk 4]
```diff
@@ -21,6 +22,48 @@ void ColumnMapping::setupByHeader(const Block & header)
...
+
+            throw Exception(
+                            ErrorCodes::INCORRECT_DATA,
+                            "Unknown field found in format header: "
+                            "'{}' at position {}\nSet the 'input_format_skip_unknown_fields' parameter explicitly "
+                            "to ignore and proceed",
+                            name, column_indexes_for_input_fields.size());
+        }
+
+        const auto column_index = idx;
```
Comment (Avogar) [inline src/Formats/ColumnMapping.cpp]: Why do we need this variable? Can't we just use `idx`?

File: src/Core/CaseAwareBlockNameMap.h [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    /// Returns NOT_FOUND in case the column_name is not in the map
+    size_t get(std::string_view column_name) const;
+
+    size_t size() const;
+
+    /// Sets the expected size of the map
+    void setSize(size_t size);
+
+    /// Constructs the map from the given block
+    void getNamesToIndexesMap(const Block & block);
```
Comment (Avogar) [inline src/Core/CaseAwareBlockNameMap.h]: The method name is strange, it's named `get*` but doesn't return anything. Something like `initFromBlock` would be better

File: src/Core/CaseAwareBlockNameMap.h [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    /// Compares two strings, using the same method as the one used internally
+    bool equal(std::string_view left, std::string_view right) const;
+private:
+
+    /// Input mode that will be used by this object
+    const FormatSettings::InputFormatColumnMatchingCaseSensitivity mode;
+
+    std::unique_ptr<CaseSensitiveBlockNameMap> match_case;
+    std::unique_ptr<CaseInsensitiveBlockNameMap> ignore_case;
+    std::unique_ptr<AutoCaseBlockNameMap> auto_case;
```
Comment (Avogar) [inline src/Core/CaseAwareBlockNameMap.h]: Let's create a single base interface `IBlockNameMap` with all required methods and store here a single `unique_ptr<IBlockNameMap>`. In `CaseAwareBlockNameMap` constructure we will create a corresponding interface. Calling virtual methods should be as cheap as using `switch` in current approach.

File: src/Core/FormatFactorySettings.h [hunk 7]
```diff
@@ -1563,6 +1563,15 @@ Allow to write information about geo columns in parquet metadata and encode colu
     DECLARE(Bool, into_outfile_create_parent_directories, false, R"(
 Automatically create parent directories when using INTO OUTFILE if they do not already exists.
 )", 0) \
+    DECLARE(InputFormatColumnMatchingCaseSensitivity, input_format_with_names_case_insensitive_column_matching, FormatSettings::InputFormatColumnMatchingCaseSensitivity::MATCH_CASE, R"(
```
Comment (Avogar) [inline src/Core/FormatFactorySettings.h]: Let's rename setting to `input_format_column_name_matching_mode` and list all supported formats in the setting description

File: src/Core/CaseAwareBlockNameMap.h [hunk 8]
```diff
@@ -0,0 +1,51 @@
+#pragma once
+
+#include <cctype>
```
Comment (Avogar) [inline src/Core/CaseAwareBlockNameMap.h]: Is this include actually needed?

File: src/Core/FormatFactorySettings.h [hunk 9]
```diff
@@ -1563,6 +1563,15 @@ Allow to write information about geo columns in parquet metadata and encode colu
...
+Defines the case sensitive mode when ingesting data through the various format containing column names (e.g., JSONEachRow).
+Supported modes:
+    - MATCH_CASE: match case-sensitively
+    - IGNORE_CASE: match case-insensitively
+    - AUTO: first tries to match case-sensitively, if fails, tries to match case-insensitively.
+
+   default value: MATCH_CASE
```
Comment (Avogar) [inline src/Core/FormatFactorySettings.h]: When we generate settings docs we add line about default value automatically, so this line is not needed

File: src/Processors/Formats/Impl/JSONEachRowRowInputFormat.h [hunk 10]
```diff
@@ -74,10 +74,12 @@ class JSONEachRowRowInputFormat : public IRowInputFormat
     /// for row like {..., "non-nullable column name" : null, ...}
 
     /// Hash table match `field name -> position in the block`. NOTE You can use perfect hash map.
-    BlockNameMap name_map;
+    CaseAwareBlockNameMap name_map;
+    /// Marks columns already seen before for this iteration
+    std::vector<bool> column_seen_before;
```
Comment (Avogar) [inline src/Processors/Formats/Impl/JSONEachRowRowInputFormat.h]: We already have `seen_columns` vector, why do we need new one? How it differs from `seen_columns` usage?

**Review by Avogar — COMMENTED** [review]
The approach looks good to me. Left a few comments.

File: src/Core/FormatFactorySettings.h (hunk 7 above)
Comment (Avogar) [reply src/Core/FormatFactorySettings.h]: Also please add new setting to SettingsChangesHistory.cpp into 25.3 version section

File: src/Core/CaseAwareBlockNameMap.cpp (hunk 1 above)
Comment (Manerone) [reply src/Core/CaseAwareBlockNameMap.cpp]: Initially this was my idea as well, but `clickhouse-gh` found an issue about it: https://github.com/ClickHouse/ClickHouse/pull/99346#discussion_r2926720504

I added a test to check this: tests/queries/0_stateless/04027_json_input_format.sql:27

File: src/Core/CaseAwareBlockNameMap.h (hunk 8 above)
Comment (Manerone) [reply src/Core/CaseAwareBlockNameMap.h]: Removed

File: src/Core/CaseAwareBlockNameMap.h (hunk 5 above)
Comment (Manerone) [reply src/Core/CaseAwareBlockNameMap.h]: Changed

File: src/Formats/ColumnMapping.cpp (hunk 4 above)
Comment (Manerone) [reply src/Formats/ColumnMapping.cpp]: Removed

File: src/Core/CaseAwareBlockNameMap.cpp (hunk 1 above)
Comment (Avogar) [reply src/Core/CaseAwareBlockNameMap.cpp]: I see the issue, good, thanks

File: src/Processors/Formats/Impl/JSONEachRowRowInputFormat.h (hunk 10 above)
Comment (Manerone) [reply src/Processors/Formats/Impl/JSONEachRowRowInputFormat.h]: Indeed, we already have one. Removing the unnecessary one that I created.

File: src/Core/CaseAwareBlockNameMap.cpp (hunk 2 above)
Comment (Manerone) [reply src/Core/CaseAwareBlockNameMap.cpp]: Changed

File: src/Core/CaseAwareBlockNameMap.cpp (hunk 3 above)
Comment (Manerone) [reply src/Core/CaseAwareBlockNameMap.cpp]: Changed

File: src/Core/CaseAwareBlockNameMap.h (hunk 6 above)
Comment (Manerone) [reply src/Core/CaseAwareBlockNameMap.h]: Changed

File: src/Core/FormatFactorySettings.h (hunk 9 above)
Comment (Manerone) [reply src/Core/FormatFactorySettings.h]: Removed

File: src/Core/FormatFactorySettings.h (hunk 7 above)
Comment (Manerone) [reply src/Core/FormatFactorySettings.h]: Changed name and description.

Added to SettingsChangesHistory.cpp I am a bit unsure about what to put in `.previous_value` and `.new_value`.

File: src/Core/CaseAwareBlockNameMap.h [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+#include <string_view>
+#include <Formats/FormatSettings.h>
+
+namespace DB
+{
+
+class Block;
+
+/// Interface for BlockNameMaps
+class IBlockNameMap {
```
Comment (rschu1ze) [inline src/Core/CaseAwareBlockNameMap.h]: (probably move this class declaration into the source file since it's not supposed to be used externally?)

File: src/Core/CaseAwareBlockNameMap.h [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+#include <string_view>
+#include <Formats/FormatSettings.h>
+
+namespace DB
+{
+
+class Block;
+
+/// Interface for BlockNameMaps
+class IBlockNameMap
```
Comment (rschu1ze) [inline src/Core/CaseAwareBlockNameMap.h]: Probably move this class declaration into the source file since it's not supposed to be used externally?

Comment (Manerone) [reply src/Core/CaseAwareBlockNameMap.h]: True, moved the class to the source file

File: src/Core/CaseAwareBlockNameMap.cpp (hunk 3 above)
Comment (Manerone) [reply src/Core/CaseAwareBlockNameMap.cpp]: This change gives us a compilation error: 
```
error: result of a `data()` call may not be null terminated, provide size information to the callee to prevent potential issues [bugprone-suspicious-stringview-data-usage,-warnings-as-errors]
```

Changing to this for now to avoid the error:

```cpp
        for(size_t i = 0; i < left.size(); i++){
            if(Poco::Ascii::toLower(left[i]) != Poco::Ascii::toLower(right[i]))
                return false;
        }
        return true;
```

We probably lose on any optimizations that `strncasecmp` could have (e.g. vectorization or SIMD)

File: src/Core/FormatFactorySettings.h [hunk 13]
```diff
@@ -1563,6 +1563,13 @@ Allow to write information about geo columns in parquet metadata and encode colu
     DECLARE(Bool, into_outfile_create_parent_directories, false, R"(
 Automatically create parent directories when using INTO OUTFILE if they do not already exists.
 )", 0) \
+    DECLARE(InputFormatColumnMatchingCaseSensitivity, input_format_column_name_matching_mode, FormatSettings::InputFormatColumnMatchingCaseSensitivity::MATCH_CASE, R"(
+Defines the column name matching mode when ingesting data through various formats (JSONEachRow, CSVWithNames, JSONColumns, BSONEachRow, RowBinaryWithNames).
+Supported modes:
+    - MATCH_CASE: match case-sensitively
+    - IGNORE_CASE: match case-insensitively
```
[the remark this answers is not in the document]
Comment (Manerone) [reply src/Core/FormatFactorySettings.h]: Name was changed according to reviewer comment

File: src/Core/SettingsChangesHistory.cpp [hunk 14]
```diff
@@ -481,6 +481,7 @@ const VersionToSettingsChangesMap & getSettingsChangesHistory()
             {"use_query_condition_cache", false, false, "New setting."},
             {"parallel_replicas_for_cluster_engines", false, true, "New setting."},
             {"parallel_hash_join_threshold", 0, 0, "New setting"},
+            {"input_format_column_name_matching_mode", "match_case", "match_case", "New setting."},
```
Comment (Avogar) [inline src/Core/SettingsChangesHistory.cpp]: It was accidentially added to 25.3 version, but current version is 26.3. Please, move it to 26.3 section

File: src/Core/CaseAwareBlockNameMap.cpp [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+    virtual bool stringCompare(std::string_view left, std::string_view right) const = 0;
+};
+
+struct CaseInsensitiveHash
+{
+    size_t operator()(const std::string_view key) const
+    {
+        SipHash hash;
+        for (char c : key)
+            hash.update(static_cast<unsigned char>(std::tolower(static_cast<unsigned char>(c))));
```
Comment (Avogar) [inline src/Core/CaseAwareBlockNameMap.cpp]: I see that `CaseInsensitiveHash` uses `std::tolower` now but `CaseInsensitiveEquality` uses `Poco::Ascii::toLower`. It's inconsistent and potentially may lead to problems if these 2 functions work differently

File: src/Formats/ColumnMapping.cpp [hunk 16]
```diff
@@ -21,6 +22,46 @@ void ColumnMapping::setupByHeader(const Block & header)
         column_indexes_for_input_fields[i] = i;
 }
 
+void ColumnMapping::addColumns(
```
Comment (Avogar) [inline src/Formats/ColumnMapping.cpp]: This method seems almost identical to `addColumns` below with `BlockNameMap`. Can we unify them? Maybe making it template?

File: src/Core/SettingsChangesHistory.cpp (hunk 14 above)
Comment (Manerone) [reply src/Core/SettingsChangesHistory.cpp]: Fixed it

File: src/Core/CaseAwareBlockNameMap.cpp (hunk 15 above)
Comment (Manerone) [reply src/Core/CaseAwareBlockNameMap.cpp]: Changed to `Poco::Ascii::toLower` in both.

File: src/Formats/ColumnMapping.cpp (hunk 16 above)
Comment (Manerone) [reply src/Formats/ColumnMapping.cpp]: Moved the similar parts to a templated function within the source file. I am not sure if there is a better way to do this, I find it a bit ugly due to the number of parameters, but it gets the job done.

I did not unify the methods since the API of `CaseAwareBlockNameMap` and `BlockNameMap` don't match, but you will see in the new method that we only have to consider the way we search for the name, so we can get by returning an optional.

File: src/Core/CaseAwareBlockNameMap.cpp [hunk 17]
```diff
@@ -0,0 +1,310 @@
...
+
+    void setSize(size_t size)
+    {
+        expected_size = size;
+        map.resize(size);
+    }
+
+    void add(const std::string_view key, size_t idx)
+    {
+        if (map.find(key) != map.end())
```
[the remark this answers is not in the document]
Comment (Avogar) [reply src/Core/CaseAwareBlockNameMap.cpp]: This concern sounds valid to me

File: src/Formats/ColumnMapping.cpp [hunk 18]
```diff
@@ -62,6 +65,62 @@ void ColumnMapping::addColumns(
             not_presented_columns.push_back(i);
     }
 }
+}
+
+namespace ErrorCodes
+{
+extern const int INCORRECT_DATA;
+}
```
Comment (Avogar) [inline src/Formats/ColumnMapping.cpp]: It's a duplicate code, we already have namespace ErrorCodes with INCORRECT_DATA above.

File: src/Core/FormatFactorySettings.h [hunk 19]
```diff
@@ -1563,6 +1563,13 @@ Allow to write information about geo columns in parquet metadata and encode colu
     DECLARE(Bool, into_outfile_create_parent_directories, false, R"(
 Automatically create parent directories when using INTO OUTFILE if they do not already exists.
 )", 0) \
+    DECLARE(InputFormatColumnMatchingCaseSensitivity, input_format_column_name_matching_mode, FormatSettings::InputFormatColumnMatchingCaseSensitivity::MATCH_CASE, R"(
+Defines the column name matching mode when ingesting data through various formats (JSONEachRow, CSVWithNames, JSONColumns, BSONEachRow, RowBinaryWithNames).
```
Comment (Avogar) [inline src/Core/FormatFactorySettings.h]:
```suggestion
Defines the column name matching mode when ingesting data through various formats (including but not limited to JSONEachRow, CSVWithNames, JSONColumns, BSONEachRow, RowBinaryWithNames).
```

It's not the complete list, so we need to make a note

File: src/Core/SettingsChangesHistory.cpp [hunk 20]
```diff
@@ -57,6 +57,7 @@ const VersionToSettingsChangesMap & getSettingsChangesHistory()
             {"mysql_datatypes_support_level", "", "decimal,datetime64,date2Date32", "Enable modern MySQL type mappings by default."},
             {"allow_experimental_json_lazy_type_hints", false, false, "New experimental setting for lazy JSON type hints"},
             {"functions_h3_default_if_invalid", true, false, "A new setting for legacy behaviour to allow invalid inputs to h3 functions"},
+            {"input_format_column_name_matching_mode", "match_case", "match_case", "New setting."},
```
Comment (Avogar) [inline src/Core/SettingsChangesHistory.cpp]: We already separated 26.3 branch, so we have to move it to 26.4 now

File: src/Core/CaseAwareBlockNameMap.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (Manerone) [reply src/Core/CaseAwareBlockNameMap.cpp]: I will still work on it, I just got busy with some other things.

File: src/Core/FormatFactorySettings.h (hunk 19 above)
Comment (Manerone) [reply src/Core/FormatFactorySettings.h]: Fixed

File: src/Core/SettingsChangesHistory.cpp (hunk 20 above)
Comment (Manerone) [reply src/Core/SettingsChangesHistory.cpp]: Fixed

File: src/Formats/ColumnMapping.cpp (hunk 18 above)
Comment (Manerone) [reply src/Formats/ColumnMapping.cpp]: Fixed

File: src/Core/CaseAwareBlockNameMap.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (Manerone) [reply src/Core/CaseAwareBlockNameMap.cpp]: I added this fix and related tests, the only part I am not so sure about is in the following scenario:

```sql
CREATE TABLE json_test (id Int, user Tuple(age Int), USER Tuple(name String));
SET input_format_column_name_matching_mode='auto';
INSERT INTO json_test FORMAT JSONEachRow {"ID": 0, "user": {"AGE": 21}, "USER": {"NAME": 401}};
SELECT * FROM json_test;

0      (0,0)   ('')
```

It seems to me that the nested fields (AGE, NAME) are always case sensitive, and this match is done via `ISerialization` interface. Is this something we want to touch as well?

File: src/Core/FormatFactorySettings.h [hunk 21]
```diff
@@ -1566,6 +1566,13 @@ Allow to write information about geo columns in parquet metadata and encode colu
     DECLARE(Bool, into_outfile_create_parent_directories, false, R"(
 Automatically create parent directories when using INTO OUTFILE if they do not already exists.
 )", 0) \
+    DECLARE(InputFormatColumnMatchingCaseSensitivity, input_format_column_name_matching_mode, FormatSettings::InputFormatColumnMatchingCaseSensitivity::MATCH_CASE, R"(
+Defines the column name matching mode when ingesting data through various formats (including but not limited to JSONEachRow, CSVWithNames, JSONColumns, BSONEachRow, RowBinaryWithNames).
+Supported modes:
+    - MATCH_CASE: match case-sensitively
```
[the remark this answers is not in the document]
Comment (Manerone) [reply src/Core/FormatFactorySettings.h]: Fixed

File: src/Processors/Formats/Impl/JSONEachRowRowInputFormat.cpp [hunk 22]
```diff
@@ -70,21 +76,19 @@ inline size_t JSONEachRowRowInputFormat::columnIndex(std::string_view name, size
...
+    if (prev_positions.size() > key_index && prev_positions[key_index].second != NOT_INITIALIZED
+        && name_map.equal(name, prev_positions[key_index].first))
     {
-        return prev_positions[key_index]->second;
+        return prev_positions[key_index].second;
     }
 
-    const auto it = name_map.find(name);
-    if (it != name_map.end())
+    auto position = name_map.get(name);
```
[the remark this answers is not in the document]
Comment (Manerone) [reply src/Processors/Formats/Impl/JSONEachRowRowInputFormat.cpp:85]: There is a regression test for this, both for `auto` and `ignore_case`:

https://github.com/ClickHouse/ClickHouse/pull/99346/changes#diff-cc605ea5f5b83a6eb6339c4a04d32d0865e169edc16653a295d69a53c5e1202eR77-R107

The second ocurrence of the key is ignored if `SET input_format_json_ignore_unnecessary_fields=true`, and ambiguity is checked if `SET input_format_json_ignore_unnecessary_fields=false`.

**Review by Avogar — APPROVED** [review]

File: src/Core/CaseAwareBlockNameMap.h [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+    /// Sets the expected size of the map
+    void setSize(size_t size);
+
+    /// Constructs the map from the given block
+    void initFromBlock(const Block & block);
+
+    /// Compares two strings, using the same method as the one used internally
+    bool equal(std::string_view left, std::string_view right) const;
+private:
+    std::unique_ptr<IBlockNameMap> map;
```
[the remark this answers is not in the document]
Comment (Manerone) [reply src/Core/CaseAwareBlockNameMap.h:41]: Fixed

**Comment by Avogar** [comment]
@Manerone I downloaded the build from one of the commits, and this query just don't throw any client or server error:
```sql
avogar-dev :) CREATE TABLE json_test (age Int, AGE Int);

CREATE TABLE json_test
(
    `age` Int,
    `AGE` Int
)

Query id: acfecf28-12a4-4ef7-921b-f41534385db7

Ok.

0 rows in set. Elapsed: 0.003 sec.

avogar-dev :) INSERT INTO json_test FORMAT JSONEachRow {"AgE": 1, "aGe": 20};

INSERT INTO json_test FORMAT JSONEachRow

Query id: 2f9829fc-cef0-443f-b31d-1c5bf0568555

Ok.

1 row in set. Elapsed: 0.004 sec.

avogar-dev :) SELECT * FROM json_test

SELECT *
FROM json_test

Query id: 0ea65591-136b-4fb0-8462-c45247bb51be

   ┌─age─┬─AGE─┐
1. │   0 │   0 │
   └─────┴─────┘

1 row in set. Elapsed: 0.003 sec.
```

So this is the issue, as I understand, you expect it to throw. But it doesn't.

**Comment by Manerone** [comment]
@Avogar just to be sure, did you set `input_format_column_name_matching_mode` to `auto` or `ignore_case`? I think if it stays in the default option (`match_case`) it will not find any matches and use the default value of the columns.

**Comment by Avogar** [comment]
Ah, right sorry, forgot to change the setting 🤦‍♂️

The actual reason is randomized setting `async_insert` in the tests. With async inserts errors don't happen on client, but are collected somewhere on server. You need to use `clientError` and add `-- Tags: no-async-insert` as the first line in the test.

**Comment by Manerone** [comment]
@Avogar thanks a lot for the help! I have applied the changes, lets see if it finally passes now.

File: tests/queries/0_stateless/04028_csv_input_format.sh [hunk 24]
```diff
@@ -0,0 +1,112 @@
...
+                       INSERT INTO test FROM INFILE '$CURDIR/data_csv/csv_with_names.csv' FORMAT CSVWithNames;"
+$CLICKHOUSE_CLIENT -q "SELECT * FROM test"
+$CLICKHOUSE_CLIENT -q "DROP TABLE test"
+
+# Test ambiguity for automatic column name matching
+$CLICKHOUSE_CLIENT -q "DROP TABLE IF EXISTS test"
+$CLICKHOUSE_CLIENT -q "CREATE TABLE test (id Int, iD Int, name String, NAME String)"
+$CLICKHOUSE_CLIENT -q "SET input_format_csv_detect_header=true;
+                       SET input_format_column_name_matching_mode='auto';
+                       INSERT INTO test FROM INFILE '$CURDIR/data_csv/csv_with_names.csv' FORMAT CSVWithNames; -- { clientError 117 }"
```
[the remark this answers is not in the document]
Comment (Manerone) [reply tests/queries/0_stateless/04028_csv_input_format.sh:23]: I don't see `run_with_error` being used anywhere in the tests. I actually see the shell tests using the same pattern as me. Example files:

- tests/queries/0_stateless/02878_use_structure_from_insertion_table_with_explicit_insert_columns.sh
- tests/queries/0_stateless/02932_refreshable_materialized_views_1.sh
- tests/queries/0_stateless/03410_variant_bad_binary_data.sh

**Comment by Avogar** [comment]
The CI is not in a good shape right now. We are working on reducing flaky failures. I will postpone merging it for a few days until the PRs with CI fixes are merged.
